http-server/src/server.rs

```diff
@@ -384,9 +439,185 @@ impl Future for ServerHandle {
+			Some(origin) => origin,
+			None => return Ok(malformed()),
+		};
+		let maybe_origin = http_helpers::read_header_value(request.headers(), "origin");
+
+		if let Err(e) = acl.verify_host(host) {
+			tracing::warn!("Denied request: {:?}", e);
+			return Ok(response::host_not_allowed());
+		}
+
+		if let Err(e) = acl.verify_origin(maybe_origin, host) {
+			tracing::warn!("Denied request: {:?}", e);
+			return Ok(response::invalid_allow_origin());
+		}
+
+		if let Err(e) = acl.verify_headers(keys, cors_request_headers) {
+			tracing::warn!("Denied request: {:?}", e);
+			return Ok(response::invalid_allow_headers());
+		}
+
+		// Only `POST` and `OPTIONS` methods are allowed.
+		match *request.method() {
+			// An OPTIONS request is a CORS preflight request. We've done our access check
+			// above so we just need to tell the browser that the request is OK.
+			Method::OPTIONS => {
+				let origin = match maybe_origin {
+					Some(origin) => origin,
+					None => return Ok(malformed()),
+				};
+
+				let allowed_headers = acl.allowed_headers().to_cors_header_value();
+				let allowed_header_bytes = allowed_headers.as_bytes();
+
+				let res = hyper::Response::builder()
+					.header("access-control-allow-origin", origin)
+					.header("access-control-allow-methods", "POST")
+					.header("access-control-allow-headers", allowed_header_bytes)
+					.body(hyper::Body::empty())
+					.unwrap_or_else(|e| {
+						tracing::error!("Error forming preflight response: {}", e);
+						internal_error()
+					});
+
+				Ok(res)
+			}
+			// The actual request. If it's a CORS request we need to remember to add
+			// the access-control-allow-origin header (despite preflight) to allow it
+			// to be read in a browser.
+			Method::POST if content_type_is_json(&request) => {
+				let origin = return_origin_if_different_from_host(request.headers()).cloned();
+				let mut res = process_validated_request(ProcessValidatedRequest {
+					request,
+					logger,
+					methods,
+					resources,
+					max_request_body_size,
+					max_response_body_size,
+					max_log_length,
+					batch_requests_supported,
+					request_start,
+				})
+				.await?;
+
+				if let Some(origin) = origin {
+					res.headers_mut().insert("access-control-allow-origin", origin);
+				}
+				Ok(res)
+			}
+			Method::GET => match health_api.as_ref() {
+				Some(health) if health.path.as_str() == request.uri().path() => {
+					process_health_request(
+						health,
+						logger,
+						methods,
+						max_response_body_size,
+						request_start,
+						max_log_length,
+					)
+					.await
+				}
+				_ => Ok(response::method_not_allowed()),
+			},
+			// Error scenarios:
+			Method::POST => Ok(response::unsupported_content_type()),
+			_ => Ok(response::method_not_allowed()),
+		}
+	}
+}
+
+/// JsonRPSee service compatible with `tower`.
+///
+/// # Note
+/// This is similar to [`hyper::service::service_fn`].
+#[derive(Debug)]
+pub struct TowerService<L> {
+	inner: ServiceData<L>,
```

Comment:
Is there a benefit to having this inner `ServiceData<L>` as well as the outer `TowerService<L>` instead of just having `TowerService<L>` have all of the props etc from `ServiceData` in it?

Reply:
orphan rule?

Reply:
Look below it implements `hyper::service::Service<hyper::Request<hyper::Body>>` which can't be done because both the trait and type are not defined by "us".

So one need the `newtype` to implement that trait AFAIU.